nemo/collections/tts/models/spectrogram_enhancer.py

```diff
@@ -0,0 +1,382 @@
+
+        # denormalize if needed, mask and remove padding
+        if normalize:
+            enhanced_spectrograms = self.unnormalize_spectrograms(enhanced_spectrograms)
+        enhanced_spectrograms = mask(enhanced_spectrograms, lengths)
+        enhanced_spectrograms = enhanced_spectrograms[:, :, :, :max_length]
+
+        return enhanced_spectrograms
+
+    def prepare_batch(self, batch):
+        if not isinstance(self.spectrogram_model, FastPitchModel):
+            raise NotImplementedError(
+                f"{type(self.spectrogram_model)} is not supported, please implement batch preparation for this model."
+            )
+
+        attn_prior, durs, speaker = None, None, None
+        (audio, audio_lens, text, text_lens, attn_prior, pitch, _, speaker,) = batch
+        mels, mel_lens = self.spectrogram_model.preprocessor(input_signal=audio, length=audio_lens)
+
+        mels_pred, *_, pitch = self.spectrogram_model(
+            text=text,
+            durs=durs,
+            pitch=pitch,
+            speaker=speaker,
+            pace=1.0,
+            spec=mels if self.spectrogram_model.learn_alignment else None,
+            attn_prior=attn_prior,
+            mel_lens=mel_lens,
+            input_lens=text_lens,
+        )
+
+        mels = mels.cpu()
+        mel_lens = mel_lens.cpu()
+        mels_pred = mels_pred.cpu()
+
+        batch_size, *_ = mels.shape
+
+        targets = []
+        conditions = []
+        lengths = []
+
+        for i in range(batch_size):
+            l = mel_lens[i].item()
+            lengths.append(l)
+
+            targets.append(mels[i, :, :l].T)
+            conditions.append(mels_pred[i, :, :l].T)
+
+        target = nn.utils.rnn.pad_sequence(targets, batch_first=True)
+        condition = nn.utils.rnn.pad_sequence(conditions, batch_first=True)
+        lengths = torch.LongTensor(lengths).unsqueeze(-1)
+
+        target = rearrange(target, "b l c -> b 1 c l")
+        condition = rearrange(condition, "b l c -> b 1 c l")
+
+        return self.move_to_correct_device((target, condition, lengths))
+
+    def training_step(self, batch, batch_idx, optimizer_idx):
+        target, condition, lengths = self.prepare_batch(batch)
+
+        with torch.no_grad():
+            target = self.normalize_spectrograms(target)
+            condition = self.normalize_spectrograms(condition)
+
+            target = mask(target, lengths)
+            condition = mask(condition, lengths)
+
+        # train discriminator
+        if optimizer_idx == 0:
+            enhanced_spectrograms = self.forward(condition, lengths, mixing=True, normalize=False)
+            fake_logits = self.D(enhanced_spectrograms, condition, lengths)
+
+            real_images = target.requires_grad_()
+            real_logits = self.D(real_images, condition, lengths)
+            d_loss = hinge_loss(real=real_logits, fake=fake_logits)
+            self.log("d_loss", d_loss, prog_bar=True)
+
+            if batch_idx % 4 == 0:
```

Comment:
What is the logic behind this if statement? Have a comment/note explaining this. Or have a method with the proper name. Like:
```
def if_<condition_name>(self, batch_idx):
    return batch_idx % 4 == 0

if self.if_<condition_name>(batch_idx):
   ...
```

Reply:
`"gradient_penalty_loss_every_n_steps"`